File: canton-3x/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/ledger/api/messages/command/submission/SubmitReassignmentRequest.scala
// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.canton.ledger.api.messages.command.submission

import com.daml.lf.data.{Ref, Time}
import com.daml.lf.value.Value
import com.digitalasset.canton.protocol.{SourceDomainId, TargetDomainId}

final case class SubmitReassignmentRequest(
    submitter: Ref.Party,
    applicationId: Ref.ApplicationId,
    commandId: Ref.CommandId,
    submissionId: Ref.SubmissionId,
    workflowId: Option[Ref.WorkflowId],
    reassignmentCommand: Either[AssignCommand, UnassignCommand],
)

final case class UnassignCommand(
    sourceDomainId: SourceDomainId,
    targetDomainId: TargetDomainId,
    contractId: Value.ContractId,
)
final case class AssignCommand(
    sourceDomainId: SourceDomainId,
    targetDomainId: TargetDomainId,
    unassignId: Time.Timestamp,
)

File: canton/community/bindings-java/src/main/java/com/daml/ledger/javaapi/data/SubmitReassignmentRequest.java
// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates.
// Proprietary code. All rights reserved.

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v2.CommandSubmissionServiceOuterClass;
import org.checkerframework.checker.nullness.qual.NonNull;

public final class SubmitReassignmentRequest {

  public static CommandSubmissionServiceOuterClass.SubmitReassignmentRequest toProto(
      @NonNull ReassignmentCommand command) {
    return CommandSubmissionServiceOuterClass.SubmitReassignmentRequest.newBuilder()
        .setReassignmentCommand(command.toProto())
        .build();
  }

  public static ReassignmentCommand fromProto(
      CommandSubmissionServiceOuterClass.@NonNull SubmitReassignmentRequest request) {
    return ReassignmentCommand.fromProto(request.getReassignmentCommand());
  }
}
